File: c++/leetcode/Check If a String Is a Valid Sequence from Root to Leaves Path in a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  bool valid(TreeNode *, vector<int> &, int);
  bool isValidSequence(TreeNode *root, vector<int> &arr)
  {
    return valid(root, arr, 0);
  }
};

bool Solution::valid(TreeNode *root, vector<int> &arr, int pos)
{
  if (!root || pos >= arr.size())
    return false;

  if (!root->left && !root->right)
  {
    if (pos == arr.size() - 1 && arr[pos] == root->val)
      return true;
    return false;
  }

  if (arr[pos] == root->val)
  {
    ++pos;
    return valid(root->left, arr, pos) || valid(root->right, arr, pos);
  }

  return false;
}

File: javascript/leetcode/Check If a String Is a Valid Sequence from Root to Leaves Path in a Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number[]} arr
 * @return {boolean}
 */
var isValidSequence = function (root, arr) {
  return (function valid(root, pos) {
    if (!root || pos >= arr.length) return false;

    if (!root.left && !root.right) {
      if (pos === arr.length - 1 && arr[pos] === root.val) return true;

      return false;
    }

    if (arr[pos] === root.val) {
      ++pos;
      return valid(root.left, pos) || valid(root.right, pos);
    }

    return false;
  })(root, 0);
};
